### ISSUE TITLE
with_huffman_tree quirk

### ISSUE THREAD

**Issue by JeremyRubin** [rust-bitcoin/rust-bitcoin#906]
One interesting issue is that the top-level key for the with_huffman_tree really should be extracted from the leaf nodes as the highest likelihood single key path, not an additional chosen key externally. Our huffman algo should be better at picking a top-level key than we are!

Ideally, I'd think the API would be something like:
```rust
with_huffman_tree(secp, scripts, missing_behavior:MB);

enum MB {
   Fail,
   UnsafeUseThisNUMSPoint(UnTweaked),
   SecureDeriveFrom(&[u8]),
}
impl Default for MB {
    fn default() {
        MB::SecureDeriveFrom("standard taproot NUMS".as_bytes())
    }
}
```

**Comment by sanket1729** [comment]
We do something similar in the rust-miniscript taproot descriptor compiler. This will get complicated as it will involve semantically understanding the scripts to infer an internal key. We need to parse those <pk> Checksig or `or(pk,other_script)`.

**Comment by JeremyRubin** [comment]
i see your point on this.

still, matching for <pk> CHECKSIG *only* for the scripts would be pretty cheap (just checking the first 2 bytes and length?).

iterating as a breadth first search should give us things as a priority queue i think, as well.

and then we could offer an third behavior in MB, which would be "always this key" to override searching the huffman tree for a key candidate.


Another option would be, in Miniscript, to just provide a function which allows extracting the most likely single signer key and returns a tree that is modified to not have that key and the key itself. The only issue i see with this approach is that:

would result in:

path one: or(pk(A)@1000000000000,pk(B)@1)
....
path N: pk(C)@1000

extracting pk(A), yielding

path one: pk(B)@1
path N: pk(C)@1000

and then path N would be higher than path one after the extraction of that path.

So really, the extraction of key has to happen before the huffman construction for correctness.

**Comment by apoelstra** [comment]
I don't think this belongs in rust-bitcoin. Yes, it is "only" checking for the CHECKSIG script but I'd rather not have any template matching in rust-bitcoin's logic. That is what rust-miniscript is for, and in fact de-templating rust-bitcoin was (my) original motivation for Miniscript.

**Comment by JeremyRubin** [comment]
i can see that logic, but then...

i'd argue that with_huffman then doesn't belong in rust-bitcoin _either_ (which is an OK outcome), it's more of a miniscript esque concern.

**Comment by apoelstra** [comment]
`with_huffman` allows you to create an arbitrary taptree with an arbitrary set of scripts. No templating.

**Comment by tcharding** [comment]
Are you planning on pursuing this issue @JeremyRubin or can we close it?

**Comment by JeremyRubin** [comment]
Leave it open :) will get around to it eventually
